File: src/main/java/challenge/dto/Recognition.java
package challenge.dto;

public class Recognition {

    private String fromUserId;
    private String toUserId;
    private String type;
    private String comment;

    public Recognition() {
    }

    public Recognition(String fromUserId, String toUserId, String type, String comment) {
        this.fromUserId = fromUserId;
        this.toUserId = toUserId;
        this.type = type;
        this.comment = comment;
    }

    public String getFromUserId() {
        return fromUserId;
    }

    public void setFromUserId(String fromUserId) {
        this.fromUserId = fromUserId;
    }

    public String getToUserId() {
        return toUserId;
    }

    public void setToUserId(String toUserId) {
        this.toUserId = toUserId;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getComment() {
        return comment;
    }

    public void setComment(String comment) {
        this.comment = comment;
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Recognition that = (Recognition) o;

        if (comment != null ? !comment.equals(that.comment) : that.comment != null) return false;
        if (fromUserId != null ? !fromUserId.equals(that.fromUserId) : that.fromUserId != null) return false;
        if (toUserId != null ? !toUserId.equals(that.toUserId) : that.toUserId != null) return false;
        if (type != null ? !type.equals(that.type) : that.type != null) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = fromUserId != null ? fromUserId.hashCode() : 0;
        result = 31 * result + (toUserId != null ? toUserId.hashCode() : 0);
        result = 31 * result + (type != null ? type.hashCode() : 0);
        result = 31 * result + (comment != null ? comment.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "Recognition{" +
                "fromUserId='" + fromUserId + '\'' +
                ", toUserId='" + toUserId + '\'' +
                ", type='" + type + '\'' +
                ", comment='" + comment + '\'' +
                '}';
    }
}

File: ui/app/models/recognition/recognition.ts
import {User} from "../user/user";
export class Recognition {
  private fromUserId: number;
  private toUserId: number;
  private type: string;
  private comment: string;
  private toUser: User;
  private fromUser: User;
  private timestamp: string;

  constructor(data: any) {
    this.fromUserId = data.fromUserId;
    this.toUserId = data.toUserId;
    this.type = data.type;
    this.comment = data.comment;
    this.toUser = data.toUser || null;
    this.fromUser = data.fromUser || null;
    this.timestamp = data.timestamp || ((new Date() as any).getTime());
  }

  getComment(): string {
    return this.comment;
  }

  getFromUserId(): number {
    return this.fromUserId;
  }

  getToUserId(): number {
    return this.toUserId;
  }

  getType(): string {
    return this.type;
  }

  toJson(): string {
    return JSON.stringify(this.asJson());
  }

  setToUser(user: User) {
    this.toUser = user;
  }

  getToUser(): User {
    return this.toUser;
  }

  getFromUser(): User {
    return this.fromUser;
  }

  getCreatedAt(): Date {
    return this.timestamp && new Date(this.timestamp) || new Date();
  }

  getIcon(): string {
    return Recognition.getIcon(this.getType());
  }

  setFromUser(user: User) {
    this.fromUser = user;
  }

  asJson(): any {
    return {
      fromUserId: this.getFromUserId(),
      toUserId: this.getToUserId(),
      type: this.getType(),
      comment: this.getComment()
    };
  }

  static asRecognitions(recognitions: any[]): Recognition[] {
    return recognitions.map((recognition: any) => {
      return new Recognition(recognition);
    });
  }

  static getIcon(type: string): string {
    let icon;
    switch (type) {
      case 'TEAMWORK':
        icon = 'img/team.svg';
        break;
      case 'EXPERIMENT':
        icon = 'img/brain.svg';
        break;
      case 'IMPROVEMENT':
        icon = 'img/thumbs-up.svg';
            break;
      case 'DELIVERY':
        icon = 'img/light-bulb.svg';
            break;
      default:
        icon = 'img/thumbs-up.svg';
            break;
    }
    return icon;
  }
}
